Various little cleanups

esheyw opened the pull request:
Hi, this seems like a neat module I'd like to use, but it's got a few what-I-consider-problems I've cleaned up:

- Set verified compatibility to only "12.331" and set maximum to the same
  - In v13 the UI is entirely AppV2, and the context hooks don't work exactly the same, so the module is currently non-functional there.
- Remove redundant `"use strict";`
  - This is implicit when using esmodules
- Removed reliance on jQuery
  - This will simplify AppV2 transition, for when you want to make a v13 compatible version, as AppV2 does not use jQuery
- Updated deprecated FontAwesome classes
  - `fas` -> `fa-solid`
- Updated HTML structure to include `form` tag as well as foundry-standard form styling divs
- Prefixed form input IDs for increased uniqueness
  - HTML IDs must be unique *across the entire page*, and `"new-password"` has too high a likelihood of collision
- Added `name`s to form inputs and use `FormDataExtended` in apply callback 
  - Now that we have a `form` tag, we can use the Foundry helper class to deal with data


Edit: I see I forgot to save without formatting, so this kinda nuked your indentation. Sorry about that. If that's a problem I can fix, or you can just apply your own formatting over.

esheyw committed:
-remove jQuery
-update deprecated fontawesome classes
-prefix form input IDs for increased uniqueness
-add names to form inputs and use FormDataExtended in apply callback
-remove redundant use strict
-set proper validated/maximum compatibility

RichardRobertson commented:
> * Set verified compatibility to only "12.331" and set maximum to the same
>
>     * In v13 the UI is entirely AppV2, and the context hooks don't work exactly the same, so the module is currently non-functional there.

I'm curious what you believe is non-functional. I booted up a v13 instance this morning before committing the verified version number change. It changed my password just fine several times.

> * Remove redundant `"use strict";`
>
>     * This is implicit when using esmodules

JavaScript is the bane of my existence. We can remove this if it's implied.

> * Removed reliance on jQuery
>
>     * This will simplify AppV2 transition, for when you want to make a v13 compatible version, as AppV2 does not use jQuery

I'm not sure what you mean by AppV2 since I'm not using the Foundry Application API directly. The Dialog class doesn't look like it had any API changes. Also, I didn't realize I'm using jQuery?

> * Updated deprecated FontAwesome classes
>
>     * `fas` -> `fa-solid`

No complaints there. Foundry still uses `fas` in their [examples](https://foundryvtt.com/api/classes/client.Dialog.html) so that's what I went with.

> * Updated HTML structure to include `form` tag as well as foundry-standard form styling divs

I was under the (apparently mistaken) impression that the dialog was a form since it had buttons. Whoops.

> * Prefixed form input IDs for increased uniqueness
>
>     * HTML IDs must be unique _across the entire page_, and `"new-password"` has too high a likelihood of collision

Understood and willing to defer to your experience (I am not a good webdev).

> * Added `name`s to form inputs and use `FormDataExtended` in apply callback
>
>     * Now that we have a `form` tag, we can use the Foundry helper class to deal with data

Literally the first time I've heard of this class and it's apparently been there for a while in Foundry. Their documentation is so bad for discoverability unless I want to read it like a dictionary from start to finish. Kudos for showing it to me.

> Edit: I see I forgot to save without formatting, so this kinda nuked your indentation. Sorry about that. If that's a problem I can fix, or you can just apply your own formatting over.

I think this happens to everyone who contributes online at least once. To be fair, it's not like I included an editorconfig or anything. I'll come back to formatting once everything else is settled.

Overall, this mostly just shows my inexperience with webdev and I appreciate your interest in my module and its improvement. I never expected it to get popular.

esheyw committed:
revert compatibility limitation

esheyw committed:
fix name overload

esheyw committed:
futureproof context option hook

esheyw commented:
> > * Set verified compatibility to only "12.331" and set maximum to the same
> >   
> >   * In v13 the UI is entirely AppV2, and the context hooks don't work exactly the same, so the module is currently non-functional there.
> 
> I'm curious what you believe is non-functional. I booted up a v13 instance this morning before committing the verified version number change. It changed my password just fine several times.

This was just user error. I've reverted this in the PR.

> > * Remove redundant `"use strict";`
> >   
> >   * This is implicit when using esmodules
> 
> JavaScript is the bane of my existence. We can remove this if it's implied.

[Yep](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode#strict_mode_for_modules)

> > * Removed reliance on jQuery
> >   
> >   * This will simplify AppV2 transition, for when you want to make a v13 compatible version, as AppV2 does not use jQuery
> 
> I'm not sure what you mean by AppV2 since I'm not using the Foundry Application API directly. The Dialog class doesn't look like it had any API changes. Also, I didn't realize I'm using jQuery?

The `html` params getting fed into your hook conditional and dialog callback are both jQuery objects, and `.find()` and `.attr()` are jQuery methods, not part of the native DOM api. As for the AppV2 of it all, this is actually an interesting case, as the `Players` application which manages the UI in question *already is* AppV2 on 13.333, but the `ContextMenu` class that handles entries is still passing jQuery objects instead of HTMLElements. This might not change until AppV1 is deleted entirely in v16 or 17, but just in case, added a check.

> > * Updated deprecated FontAwesome classes
> >   
> >   * `fas` -> `fa-solid`
> 
> No complaints there. Foundry still uses `fas` in their [examples](https://foundryvtt.com/api/classes/client.Dialog.html) so that's what I went with.

Yea it's not like they don't work, so Foundry's been rather slow to update, but the FA docs say deprecated so I try and avoid em

> > * Updated HTML structure to include `form` tag as well as foundry-standard form styling divs
> 
> I was under the (apparently mistaken) impression that the dialog was a form since it had buttons. Whoops.

Nope, V1 dialogs just dump your `content` into a (starting from the root element) `div > section > div.dialog-content`, gotta provide the `<form>` yourself. DialogV2 wraps user content in `<form>` by default though.

> > * Prefixed form input IDs for increased uniqueness
> >   
> >   * HTML IDs must be unique _across the entire page_, and `"new-password"` has too high a likelihood of collision
> 

Thus sayeth [the spec](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/id)

> 
> > * Added `name`s to form inputs and use `FormDataExtended` in apply callback
> >   
> >   * Now that we have a `form` tag, we can use the Foundry helper class to deal with data
> 
> Literally the first time I've heard of this class and it's apparently been there for a while in Foundry. Their documentation is so bad for discoverability unless I want to read it like a dictionary from start to finish. Kudos for showing it to me.

Their docs are not in the best state at the moment for a number of reasons, but also docs aren't tutorials, and the development guides are fairly barebones. If you're not already hanging out and asking questions in #module-development on the main Foundry discord, you should be. Fastest way to learn. The [community wiki](https://foundryvtt.wiki/en/home) has been seeing some updates in the last few months as well, and I think all the really out of date info has been archived, just check the last updated date just in case.

> > Edit: I see I forgot to save without formatting, so this kinda nuked your indentation. Sorry about that. If that's a problem I can fix, or you can just apply your own formatting over.
> 
> I think this happens to everyone who contributes online at least once. To be fair, it's not like I included an editorconfig or anything. I'll come back to formatting once everything else is settled.
> 
> Overall, this mostly just shows my inexperience with webdev and I appreciate your interest in my module and its improvement. I never expected it to get popular.

Honestly, I saw this come up in the #package-releases stream on discord and went "wait, how *do* you update a password?", looked at the code, was disappointed by "oh, just a document update, guess you shoulda read the User schema at some point", but then I saw a bunch of little things to tweak and was annoyed at the other project I was working on and... here we are.

Edit: Also don't be too down on yourself, Foundry modding is a *lot* of people's first webdev-ish project, including mine. Lots of people on the discord who only know how to code because they wanted some bit of UI nobody else had made yet. I've always felt awkward asking my players what they want their password, but never bothered to script up a solution, so this is legitimately useful (to someone, I'm sure, I don't actually have a game right now)

esheyw committed:
remove debug log

RichardRobertson commented:
> Edit: Also don't be too down on yourself, Foundry modding is a lot of people's first webdev-ish project, including mine. Lots of people on the discord who only know how to code because they wanted some bit of UI nobody else had made yet. I've always felt awkward asking my players what they want their password, but never bothered to script up a solution, so this is legitimately useful (to someone, I'm sure, I don't actually have a game right now)

Not down on myself at all. I'm down on JavaScript. Also been hating on Foundry's documentation since version 0.8 when I first started messing with it in script macros. It has not improved.

I've got quite a bit going on this weekend but I plan to dig into your proposed changes on Monday. I'll probably implement a style of some kind too.

RichardRobertson commented:
I have decided on a code style and added editorconfig, ESLint, and Prettier. This makes your pull request conflict, but I still want to integrate your changes. I'll take care of the manual work.
